api/http/management.go

```diff
@@ -408,67 +409,84 @@ func (h ManagementController) websocketWriter(
-							Type:   app.DelayMessage,
-							Offset: recordedBytes,
-							DelayMs: uint16(float64(keystrokeDelay) *
-								0.000001),
-							TerminalHeight: 0,
-							TerminalWidth:  0,
+
+						//override original message with shell error
+						forwardedMsg = userErrMsg
+					} else {
+						// session recording
+						b, e := recorderBuffered.Write(mr.Body)
+						if e != nil {
+							l.Errorf("session logging: "+
+								"recorderBuffered.Write"+
+								"(len=%d)=%d,%+v",
+								len(mr.Body), b, e)
 						}
-						n, _ := controlRecorderBuffered.Write(
-							controlMsg.MarshalBinary())
-						l.Debugf("saving control delay message: %+v/%d",
-							controlMsg, n)
-						controlBytes += n
-						if controlBytes >= app.MessageSizeLimit {
-							l.Infof("closing session session_id=%s"+
-								"on control data limit reached.",
-								session.ID)
-							//see MEN-4448
-							break Loop
+						recordedBytes += len(mr.Body)
+						session.BytesRecordedMutex.Lock()
+						session.BytesRecorded = recordedBytes
+						session.BytesRecordedMutex.Unlock()
+
+						// session control recording
+						timeNowUTC := time.Now().UTC().UnixNano()
+						keystrokeDelay := timeNowUTC - lastKeystrokeAt
+						if keystrokeDelay >= keyStrokeDelayRecordingThresholdNs {
+							if keystrokeDelay > keyStrokeMaxDelayRecording {
+								keystrokeDelay = keyStrokeMaxDelayRecording
+							}
+							controlMsg := app.Control{
+								Type:   app.DelayMessage,
+								Offset: recordedBytes,
+								DelayMs: uint16(float64(keystrokeDelay) *
+									0.000001),
+								TerminalHeight: 0,
+								TerminalWidth:  0,
+							}
+							n, _ := controlRecorderBuffered.Write(
+								controlMsg.MarshalBinary())
+							l.Debugf("saving control delay message: %+v/%d",
+								controlMsg, n)
+							controlBytes += n
+						}
+						lastKeystrokeAt = timeNowUTC
+						b, e = recorderBuffered.Write(mr.Body)
+						if e != nil {
+							l.Errorf("session logging: "+
+								"recorderBuffered.Write(len=%d)=%d,%+v",
+								len(mr.Body), b, e)
 						}
 					}
-					lastKeystrokeAt = timeNowUTC
-					b, e := recorderBuffered.Write(mr.Body)
-					if e != nil {
-						l.Errorf("session logging: "+
-							"recorderBuffered.Write(len=%d)=%d,%+v",
-							len(mr.Body), b, e)
-					}
-					recordedBytes += len(mr.Body)
-					session.BytesRecordedMutex.Lock()
-					session.BytesRecorded = recordedBytes
-					session.BytesRecordedMutex.Unlock()
-					if recordedBytes >= app.MessageSizeLimit {
-						l.Infof("closing session on limit reached.")
-						//see https://tracker.mender.io/browse/MEN-4448
-						break Loop
-					}
+
 				case shell.MessageTypeStopShell:
 					l.Debugf("session logging: recorderBuffered.Flush()"+
 						" at %d on stop shell", recordedBytes)
 					recorderBuffered.Flush()
 				}
 			}
 
-			err = conn.WriteMessage(websocket.BinaryMessage, msg.Data)
+			err = conn.WriteMessage(websocket.BinaryMessage, forwardedMsg)
```

Comment:
ok, so each message above the recording limit means sending the error message to the ws? I kind of dont like it. maybe once is enough and ignore all the following shell messages?

Reply:
yes, every message above limit means `send err to ui + send stop to device`.

I get your point and I considered this. I preferred sending it 'many times' until the device stops. my thinking was - if we already requested a stop, and it's still talking, maybe the message didn't make it after all?

I decided for it just-in-case, I'm not 100% sure if in some failure scenarios NATs doesn't drop messages. or, I'm almost sure there is some possibility of a message not making it to the device... distributed system etc.